### ISSUE TITLE
Why is map  low training on VOC？

### ISSUE THREAD

**Issue by xtanitfy** [zengarden/light_head_rcnn#50]
I trained on my own dataset that only three classes,then map is 78%.
But I trained on VOC0712 that have 20 classes,,the map is only 21%.
I guess the reason is that I only used one 1080TI, resulting in a serious under-fitting。

**Comment by xtanitfy** [comment]
More categories, fewer batches result in less precision.I don't know if my guess is correct or not.
If the number of graphics cards is small, it can be forwarded multiple times before back propagation.But the effect of batchnorm is not that big.

**Comment by aggpankaj2** [comment]
@xtanitfy have you tried with coco database. i got only 5% MAP

**Comment by xtanitfy** [comment]
@aggpankaj2 Not yet.I trained VOC again by reducing resolution and increasing batchsize to 8,but the map is only 23%.I really don't know where the problem is.

**Comment by hyw95** [comment]
i have got the same problem.i have used  one 1080ti to run test.py on coco dataset.the result mAP only 0.054.Have you find the solution?

**Comment by xtanitfy** [comment]
Maybe  it's need the same software version and the same hardware with 8 gpus to reproduce the same results.But there is also a great possibility that it will not come back. But there exists  this way to do by conditional friends.Of course,the best way is to need the author's reply.

**Comment by Yamin05114** [comment]
simply because you didn’t load the imagenet weights I think. In the report, all the feature extractors are initialized with imagenet weights while the code didn’t do it. So simply download and load weights and freeze the first 2 convs before training. I haven’t tried it but I am sure we have to do it anyways.

**Comment by xtanitfy** [comment]
@Yamin05114  After checking and debugging the code by me,i find it's not the reason that no loading imagenet weights..

**Comment by TonyTangYu** [comment]
@xtanitfy   I want to train light head rcnn on voc now but I don't know how to use the pascal_voc dataset. Do I need to transform voc format to coco ? Could you tell me how to do it or provide an instruction? Thank you!

**Comment by XingLiuJia** [comment]
> @xtanitfy I want to train light head rcnn on voc now but I don't know how to use the pascal_voc dataset. Do I need to transform voc format to coco ? Could you tell me how to do it or provide an instruction? Thank you!

Do you finish it on VOC ?

**Comment by XingLiuJia** [comment]
> @xtanitfy I want to train light head rcnn on voc now but I don't know how to use the pascal_voc dataset. Do I need to transform voc format to coco ? Could you tell me how to 


> I trained on my own dataset that only three classes,then map is 78%.
> But I trained on VOC0712 that have 20 classes,,the map is only 21%.
> I guess the reason is that I only used one 1080TI, resulting in a serious under-fitting。

Do I need to transform voc format to coco ? Could you tell me how to do it or provide an instruction? Thank you!

**Comment by xtanitfy** [comment]
@TonyTangYu  I convert the KITTI format to COCO format:
# -*- coding: utf-8 -*-

import numpy as np
import json
import os
import cv2
import shutil


DATASET_ROOT_DIR = '/disk1/DataSet/VKITTI/training'
KITTI_LABLE_DIR = DATASET_ROOT_DIR + '/label_2'
KITTI_VAL_DIR = '/disk1/DataSet/VKITTI/ImageSets/val.txt' 
KITTI_TRAIN_DIR = '/disk1/DataSet/VKITTI/ImageSets/train.txt'

OUT_FILE = 'out.txt'

class_names = ('aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus',
                       'car', 'cat', 'chair', 'cow', 'diningtable', 'dog',
                       'horse', 'motorbike', 'person', 'pottedplant', 'sheep',
                       'sofa', 'train', 'tvmonitor')

DST_ROOT_DIR = '/disk2/tf_workspace/Detection_two_stage/light_head_rcnn'

def bbox_transform_inv(bbox):
    """convert a bbox of form [xmin, ymin, xmax, ymax] to [cx, cy, w, h]. Works
    for numpy array or list of tensors.
    """

    xmin, ymin, xmax, ymax = bbox
    out_box = [[]]*4

    width       = xmax - xmin + 1.0
    height      = ymax - ymin + 1.0
    out_box[0]  = xmin + 0.5*width 
    out_box[1]  = ymin + 0.5*height
    out_box[2]  = width
    out_box[3]  = height

    return out_box
  
  
def load_kitti_from_file(anno_file_path,img_dir):
    with open(anno_file_path,'r') as f:
        lines = f.readlines()
    
    anno_file = anno_file_path.split('/')[-1]
    
    json_objs = {}
    objects = []
    sample = "a -1 -1 0.0 0 0 0 0 0.0 0.0 0.0 0.0 0.0 0.0 0.0"
    for line in lines:
        if len(line) < len(sample):
            continue
            
         
        items = line.strip().split(' ')
        one_object = {}
        one_object['box'] = bbox_transform_inv([int(items[r]) for r in range(4,8)])
        
        one_object['occ'] = 0
        tag = items[0].lower()
        
        assert tag  in class_names
        one_object['tag'] = tag
        
        is_ignore = {}
        is_ignore['ignore'] = 0
        one_object['extra'] = is_ignore
        
        objects.append(one_object)
    json_objs['gtboxes'] = objects
    name = anno_file.split('.')[0]
    json_objs['fpath'] = os.path.join('/',img_dir,str(name + '.jpg'))
    json_objs['dbName'] = 'ADAS'
    
    dbInfo = {}
    dbInfo['vID'] = 'ADAS_V0'
    dbInfo['frameID'] = -1
    json_objs['dbInfo'] = dbInfo
    
    img_path = DST_ROOT_DIR+'/data/MSCOCO'+json_objs['fpath']
    if os.path.exists(img_path) == False:
        src_img = os.path.join(DATASET_ROOT_DIR+'/image_2',name + '.jpg')
        shutil.copy(src_img,img_path)
        
    assert os.path.exists(img_path) == True,'{} not exists!'.format(img_path)
    im = cv2.imread(img_path)
    h,w,_ = [int(v) for v in im.shape]
    json_objs['width'] = w
    json_objs['height'] = h    
    json_objs['ID'] = str(anno_file.split('.')[0])+'.jpg'
    
    return json_objs

    
def get_all_files(train_val_file):
    '''
    files = os.listdir(KITTI_LABLE_DIR)
    label_files = [f for f in files if '.txt' in f]
    return label_files
    '''
    with open(train_val_file,'r') as f:
        lines = f.readlines()
       
    return  [f.strip()+'.txt' for f in lines]
    
    
 
def convert_files(files,out_file,img_dir):
    cnt = 0
    with open(out_file,'w') as f_out:
        for f in files:
            json_objs = load_kitti_from_file(os.path.join(KITTI_LABLE_DIR,f),img_dir)
            json_str = json.dumps(json_objs)
            
            f_out.write(json_str+'\n')
            print ('{}/{}'.format(cnt,len(files)))
            cnt += 1
    
if __name__ == '__main__':

    assert os.path.exists(KITTI_VAL_DIR)
    val_files = get_all_files(KITTI_VAL_DIR)
    convert_files(val_files,'coco_minival2014.odgt','val2014')

    assert os.path.exists(KITTI_TRAIN_DIR)
    train_files = get_all_files(KITTI_TRAIN_DIR)
    convert_files(train_files,'coco_trainvalmini.odgt','train2014')

**Comment by xtanitfy** [comment]
Convert  VOC format to KITTI foramt:
#!/usr/bin/python
# -*- coding: UTF-8 -*-

from xml.dom.minidom import parse
import xml.dom.minidom
import os


SRC_DIR='../voc_person/Annotations'
DST_DIR='../kitti_person/training/label_2'

CLASS_NAMES = ('aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus',
                       'car', 'cat', 'chair', 'cow', 'diningtable', 'dog',
                       'horse', 'motorbike', 'person', 'pottedplant', 'sheep',
                       'sofa', 'train', 'tvmonitor')
                       
difficult_cnt=0                      
def parse_xml_file(xmlfilename,dst_dir,dst_filename):
    if xmlfilename.split('.')[-1] != 'xml':
        print (xmlfilename + ' is not xml file')
        return 
    
    DOMTree = xml.dom.minidom.parse(xmlfilename)
    annotation = DOMTree.documentElement

    filename = xmlfilename.split('/')[-1]
    #print kitti_file
    filename = annotation.getElementsByTagName('filename')[0]
    
    size = annotation.getElementsByTagName('size')[0]
    width = size.getElementsByTagName('width')[0]
    #print ('width:%d'%int(width.childNodes[0].data))
    height = size.getElementsByTagName('height')[0]
    #print ('height:%d'%int(height.childNodes[0].data))
    depth = size.getElementsByTagName('depth')[0]
    #print ('depth:%d'%int(depth.childNodes[0].data))

    objects = annotation.getElementsByTagName('object')
    
    objs = []
    
    global difficult_cnt
    for object in objects:
        #print ('===========')
        name = object.getElementsByTagName('name')[0]
        #print "name: %s" % name.childNodes[0].data 
        
        if name.childNodes[0].data not in CLASS_NAMES:
            continue
        str = ''
        str += name.childNodes[0].data + ' '
        
        difficult =  object.getElementsByTagName('difficult')[0]
        if int(difficult.childNodes[0].data) > 0:
            difficult_cnt += 1 
            continue 
        str += difficult.childNodes[0].data + ' ' 
        bndbox = object.getElementsByTagName('bndbox')[0]

        str += '-1 0.0 '
        xmin = bndbox.getElementsByTagName('xmin')[0]
        #print "xmin: %s" % xmin.childNodes[0].data
        #center_x = 

        str += xmin.childNodes[0].data + ' ' 

        ymin = bndbox.getElementsByTagName('ymin')[0]
        #print "ymin: %s" % ymin.childNodes[0].data
        str += ymin.childNodes[0].data + ' '

        xmax = bndbox.getElementsByTagName('xmax')[0]
        #print "xmax: %s" % xmax.childNodes[0].data
        str += xmax.childNodes[0].data + ' '

        ymax = bndbox.getElementsByTagName('ymax')[0]
        #print "ymax: %s" % ymax.childNodes[0].data
        str += ymax.childNodes[0].data + ' '

        str += '0.0 0.0 0.0 0.0 0.0 0.0 0.0\n'
        
        objs.append(str)
    
    if len(objs) == 0:
        return 0
        
    kitti_file = dst_dir + '/' + dst_filename
    with open(kitti_file,"w+") as f:
        for item in objs:
            f.write(item)
        f.close()
    return len(objs)
    

if __name__ == '__main__':
    files = os.listdir(SRC_DIR)
    
    cnt = 0
    for label_f in files:
        dst_filename = label_f.split('.')[0]+'.txt'
        parse_xml_file(SRC_DIR + '/' + label_f,DST_DIR,dst_filename)
        print '{}/{}'.format(cnt,len(files))
        cnt += 1
